### ISSUE TITLE
Manage Time-Related code via `TimeProvider`  (.NET8)

### ISSUE THREAD

**Issue by MrBogomips** [FastEndpoints/FastEndpoints#458]
In order to improve the testability of the system FastEndpoints should not depend on `DateTime.UtcNow`.

It's highly advisable to introduce a System Time Provider.

**Comment by dj-nitehawk** [comment]
time provider is a new thing coming in .net 8 right?

**Comment by MrBogomips** [comment]
Yes, and the AspNet Core Team have already integrated the feature (https://github.com/dotnet/aspnetcore/issues/47472).

Just a reminder in case of subtle implications.

I've analized some parts and seems that FE relies on `DateTime.NowUtc` that should be fine.

**Comment by dj-nitehawk** [comment]
> seems that FE relies on DateTime.NowUtc that should be fine.

does that mean `DateTime.UtcNow` uses the `TimeProvider` internally after .net8?
so we no changes in FE is necessary right?

**Comment by MrBogomips** [comment]
No, for performance reason I suppose.
The actual implementation rely on a system call:
```
[System.Runtime.InteropServices.DllImportAttribute("libSystem.Native", EntryPoint = "SystemNative_GetSystemTimeAsTicks", ExactSpelling = true)]
internal static extern partial long GetSystemTimeAsTicks();
```

All time-related code should depend on `TimeProvider` (An abstract class) in order to be _testable_ with the Microsoft approach.

One of this piece, the one I stumbled upon, is the `RefreshTokenService`.

**Comment by dj-nitehawk** [comment]
alright...
will leave this issue open for now.
will start using `TimeProvider` when we upgrade the package to `.net 8.0`
any idea what's gonna happen to projects using FE on earlier .net versions?
is there any better solution than writing those nasty "multi-targetting" code blocks with `#IF NET8_0`?

**Comment by dj-nitehawk** [comment]
will look in to that when i get a chance...

**Comment by dj-nitehawk** [comment]
@MrBogomips implemented in `v5.13.0.1-beta`.
see changelog here: https://github.com/FastEndpoints/FastEndpoints/blob/main/Src/Library/changelog.md#-new

and i believe the `FastEndpoints.Security` package is only place we need to use the `TimeProvider` in, since all other places don't seem to effect what the end user can interact with.

if you find something else that requires the use of `TimeProvider`, do let me know and it will be easy to add support for that.
